### PR TITLE
Simplify function of Sum and Product on simplify

### LINKED ISSUE #14219: Sum((diag(0, 2, -3))**n, (n, 0, 3)).doit() causes an error
```
>>> k  = symbols('k', positive = True, integer = True)
>>> Sum(x**n/(n + 1), (n, 0, k)).doit()
⎧   k + 1                         log(-x + 1)                     
⎪- x     ⋅lerchphi(x, 1, k + 2) - ───────────  for │x│ ≤ 1 ∧ x ≠ 1
⎪                                      x                          
⎪                                                                 
⎪                  k                                              
⎪                 ____                                            
⎪                 ╲                                               
⎨                  ╲      n                                       
⎪                   ╲    x                                        
⎪                   ╱  ─────                        otherwise     
⎪                  ╱   n + 1                                      
⎪                 ╱                                               
⎪                 ‾‾‾‾                                            
⎪                n = 0                                            
⎩                                                                 
```
An evaluation of a finite sum containing a convergence condition

```
>>> from sympy.concrete.gosper import gosper_sum
>>> gosper_sum(x**n, (n, 0, k))
   k    
x⋅x  - 1
────────
 x - 1  
>>> gosper_sum(n*x**n, (n, 0, k))
  ⎛     k      k    k    ⎞
x⋅⎝k⋅x⋅x  - k⋅x  - x  + 1⎠
──────────────────────────
                2         
         (x - 1)          
```
The above have `x*x**k` rather than `x**(k+1)`

```
>>> A = diag(0, 2, -3)
>>> A**0 + A**1 + A**2 + A**3
⎡1  0    0 ⎤
⎢          ⎥
⎢0  15   0 ⎥
⎢          ⎥
⎣0  0   -20⎦
>>> Sum(A**n, (n, 0, 3)).doit() 
Traceback (most recent call last):

  File "<ipython-input-83-bb0cff02949f>", line 1, in <module>
    Sum(A**n, (n, 0, 3))

  File ".\sympy\core\decorators.py", line 132, in binary_op_wrapper
    return func(self, other)

  File ".\sympy\matrices\common.py", line 2052, in __pow__
    return a._matrix_pow_by_jordan_blocks(num)

  File ".\sympy\matrices\matrices.py", line 1888, in _matrix_pow_by_jordan_blocks
    jordan_cell_power(j, num)

  File ".\sympy\matrices\matrices.py", line 1873, in jordan_cell_power
    raise ValueError("Matrix det == 0; not invertible")

ValueError: Matrix det == 0; not invertible
```

```
>>> Sum((n**(-2/n))**n, (n, 1, oo)).doit()
  ∞          
_____        
╲            
 ╲          n
  ╲   ⎛ -2 ⎞ 
   ╲  ⎜ ───⎟ 
   ╱  ⎜  n ⎟ 
  ╱   ⎝n   ⎠ 
 ╱           
╱            
‾‾‾‾‾        
n = 1        
>>> Sum(n**(-2), (n, 1, oo)).doit()
 2
π 
──
6 
```

### PR DESCRIPTION
**Pull request by oscargus**
#### References to other Issues or PRs

Relates to #22260 (also see  #22264 which applies simplification as part of `doit` which also makes sense)
Closes #14219 (just adding test, fixed elsewhere)

#### Brief description of what is fixed or changed
The function in the `Sum` and `Product` is simplified when calling `simplify`.

Also, `product_simplify` is improved a bit, honoring simplify keyword arguments and using `sift`.

#### Release Notes

* concrete
   * The function of `Sum` and `Product` is now simplified when calling `simplify`.

### CODE REVIEW

**Comment by oscargus** [comment]
Seems like the problem causing the failed test was fixed elsewhere.

File: sympy/simplify/simplify.py [hunk 1]
```diff
@@ -896,34 +896,37 @@ def __refactor(val):
...
+                             lambda i: isinstance(i, Product), binary=True)
 
-    for term in terms:
-        if isinstance(term, Product):
-            p_t.append(term)
+    tmp_prod_terms = []
+    for term in prod_terms:
+        f = term.function.simplify(**kwargs)
+        if f.is_zero:
+            return S.Zero
```
Comment (oscarbenjamin) [inline sympy/simplify/simplify.py:910]: Thee probably needs to be some checking here:
```python
In [4]: p = Product(Product(sin(t)**2+cos(t)**2-1, (k, 0, 1))*Product(1/k, (k, 0, 1)), (n, 0, 1))

In [5]: p
Out[5]: 
       1                                               
─┬───────────┬─                                        
 │           │     1         1                         
 │           │  ─┬────┬─   ─┬──┬─                      
 │           │   │    │  1  │  │     2         2       
 │           │   │    │  ─⋅ │  │  sin (t) + cos (t) - 1
 │           │   │    │  k  │  │                       
 │           │   │    │    k = 0                       
 │           │   k = 0                                 
 │           │                                         
     n = 0                                             

In [6]: simplify(p)
Out[6]: 0

In [7]: p.doit()
Out[7]: 
                           4
    ⎛   2         2       ⎞ 
zoo⋅⎝sin (t) + cos (t) - 1⎠ 

In [8]: p.doit().simplify()
Out[8]: nan
```

Comment (oscargus) [reply sympy/simplify/simplify.py:910]: In which way should they evaluate? But of course it is possible to skip the early exit.

Comment (oscarbenjamin) [reply sympy/simplify/simplify.py:910]: The most important thing is that we don't get a definite-looking result for a case like this that is really undefined.

Comment (oscargus) [reply sympy/simplify/simplify.py]: I came back to this now and it turns out that the problem is actually not in this implementation (although return 0 is not a good idea).

In current master:
```
In [3]: p = Product(sin(t)**2+cos(t)**2-1, (k, 0, 1))*Product(1/k, (k, 0, 1))

In [4]: p.simplify()
Out[4]: 0

In [5]: p.doit()
Out[5]: zoo*(sin(t)**2 + cos(t)**2 - 1)**2
```

Comment (oscargus) [reply sympy/simplify/simplify.py]: Which seems to come from:
```
In [3]: p = 0*Product(1/k, (k, 0, 1))

In [4]: p
Out[4]: 0
```

Comment (oscargus) [reply sympy/simplify/simplify.py]: I think this comes from:
https://github.com/sympy/sympy/blob/a3826118a64c252bf6170a10b87d15a65728a8f4/sympy/core/mul.py#L684-L691

where especially `is_finite == False` actually should be `is_finite != True`, but that will also give some other interesting effects, like `0*x` not evaluating to `0` with default assumptions.

Comment (oscargus) [reply sympy/simplify/simplify.py]: Nah, that should rather be something like:
``` python
            # we know for sure the result will be 0 except the multiplicand
            # is infinity or a matrix
            if any(isinstance(c, MatrixExpr) for c in nc_part):
                return [coeff], nc_part, order_symbols
            if any(c.is_finite == False for c in c_part):
                return [S.NaN], [], order_symbols
            possibly_not_finite = [c for c in c_part if c.is_finite == None]
            return [coeff], possibly_not_finite, order_symbols
```

Comment (oscargus) [reply sympy/simplify/simplify.py]: Changing this leads to that the code breaks hard. Not only test failures, but test errors. Will update once everything is ran.

Comment (oscargus) [reply sympy/simplify/simplify.py]: Previous was wrong. Correct is probably:
``` python
            # we know for sure the result will be 0 except the multiplicand
            # is infinity or a matrix
            if any(isinstance(c, MatrixExpr) for c in nc_part):
                return [coeff], nc_part, order_symbols
            if any(c.is_finite == False for c in c_part):
                return [S.NaN], [], order_symbols
            c_part = [c for c in c_part if c.is_finite == None]
            if not c_part:
                return [coeff], [], order_symbols
```

But quite a lot of failures and `simplify` still evaluates incorrectly to 0.

Comment (oscarbenjamin) [reply sympy/simplify/simplify.py]: See also #17224

**Comment by oscarbenjamin** [comment]
Looks good.

**Pull request merged**